sdb/commands/stacks.py

```diff
@@ -216,86 +248,120 @@ def find_module_memory_segment(mod_name: str) -> Tuple[int, int]:
+                    "module '{:s}' doesn't exist or isn't currently loaded".
+                    format(self.args.module))
+            self.mod_end, self.mod_start = Stacks.find_mod_offset(
                 self.args.module)
-            assert mod_start != -1
-            mod_end = mod_start + mod_size
 
+    def print_header(self):
         header = "{:<18} {:<16s}".format("TASK_STRUCT", "STATE")
         if not self.args.all:
             header += " {:>6s}".format("COUNT")
         print(header)
         print("=" * 42)
 
+    #
+    # Return a boolean indicating if this stack matches the state, function or module
+    # arguments (if any) and a string representation of the stack.
+    #
+    def match_stack(self, stack_key: Tuple[Any], tasks: List[drgn.Object],
+                    print_cb: Callable[[str], None]):
+        match, stacktrace_info = True, ""
+
+        task_state = stack_key[0]
+        if self.args.tstate and self.args.tstate != task_state:
+            return False, stacktrace_info
+
+        if self.args.all:
+            for task in tasks:
+                stacktrace_info += "{:<18s} {:<16s}\n".format(
+                    hex(task.value_()), task_state)
+        else:
+            stacktrace_info += "{:<18s} {:<16s} {:6d}\n".format(
+                hex(tasks[0].value_()), task_state, len(tasks))
+        mod_match, func_match = False, False
         #
-        # We inspect and group the tasks by recording their state and
-        # stack frames once in the following loop. We do this because
-        # on live systems state can change under us, thus running
-        # something like sdb.get_prog().stack_trace(task) twice (once for
-        # grouping and once for printing) could yield different stack
-        # traces resulting into misleading output.
+        # Note on the type-check being ignored:
+        # The original `stack_key` type is a list where the first
+        # element is a string and the rest of them are integers
+        # but this is not easily expressed in mypy, thus we ignore
+        # the assignment error below.
         #
+        frame_pcs: Tuple[int] = stack_key[1:]  # type: ignore[assignment]
+        for frame_pc in frame_pcs:
+            if self.args.module and self.mod_start <= frame_pc < self.mod_end:
+                mod_match = True
+
+            try:
+                sym = sdb.get_symbol(frame_pc)
+                func, offset = sym.name, frame_pc - sym.address
+                if self.args.function and self.args.function == func:
+                    func_match = True
+            except LookupError:
+                func, offset = hex(frame_pc), 0x0
+
+            stacktrace_info += "{:18s}{}+{}\n".format("", func, hex(offset))
+        if (self.args.function and not func_match) or (self.args.module and
+                                                       not mod_match):
+            match = False
+        if match:
+            print_cb(stacktrace_info)
+        return match
```

Comment:
It seems like `match` starts as `True` and gets set to `False` only in the above if-clause. Why don't we get rid of it and do something like this:
```
if (.....):
    return False
print_cb(stacktrace_info)
return True
```